charts/scalardb-graphql/values.yaml

```diff
@@ -0,0 +1,150 @@
+# Default values for scalardb-graphql.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+
+nameOverride: ""
+fullnameOverride: ""
+
+replicaCount: 3
+
+scalarDbGraphQlConfiguration:
+  # -- The database contact point such as a hostname of Cassandra or a URL of Cosmos DB account.
+  contactPoints: cassandra
+  # -- The database port number.
+  contactPort: 9042
+  # -- The username of the database. For Cosmos DB please leave blank. For Dynamo DB please specify key id here.
+  username: cassandra
+  # -- The password of the database. For Cosmos DB, Dynamo DB please specify a secret key here.
+  password: cassandra
+  # -- Storage implementation. Either cassandra or cosmos or dynamo or jdbc can be set.
+  storage: cassandra
+  # -- The type of the transaction manager.
+  transactionManager: consensus-commit
+  # -- The log level of Scalar DB GraphQL
+  logLevel: INFO
+  # -- Path component of the URL of the GraphQL endpoint. The default is /graphql.
+  path: /graphql
+  # -- Comma-separated list of namespaces of tables for which the GraphQL server generates a schema.
+  namespaces: ""
+  # -- Whether the GraphQL server serves GraphiQL IDE. The default is true.
+  graphiql: "true"
+
+image:
+  # -- Docker image reposiory of Scalar DB GraphQL.
+  repository: ghcr.io/scalar-labs/scalardb-graphql
+  # -- Specify a image pulling policy.
+  pullPolicy: IfNotPresent
+  # -- Docker tag of the image.
+  tag: 3.5.0
+
+# -- Optionally specify an array of imagePullSecrets. Secrets must be manually created in the namespace.
+imagePullSecrets: [name: reg-docker-secrets]
+
+strategy:
+  rollingUpdate:
+    # -- The number of pods that can be created above the desired amount of pods during an update
+    maxSurge: 25%
+    # -- The number of pods that can be unavailable during the update process
+    maxUnavailable: 25%
+  # -- New pods are added gradually, and old pods are terminated gradually, e.g: Recreate or RollingUpdate
+  type: RollingUpdate
+
+service:
+  # -- service types in kubernetes.
+  type: ClusterIP
+  # -- Scalar DB GraphQL server port.
+  port: 8080
+
+ingress:
+  # -- Enable ingress resource.
+  enabled: true
+  # -- The ingress class name. Specify "alb" for AWS Application Load Balancer.
+  className: ""  # nginx / alb / ...
+  # -- The class-specific annotations for the ingress resource.
+  annotations:
+    # AWS
+    alb.ingress.kubernetes.io/scheme: internal
+    alb.ingress.kubernetes.io/target-group-attributes: stickiness.enabled=true,stickiness.lb_cookie.duration_seconds=60
+    alb.ingress.kubernetes.io/target-type: ip
+    alb.ingress.kubernetes.io/healthcheck-path: /graphql?query=%7B__typename%7D
+
+    # Azure
+    # kubernetes.io/ingress.class: azure/application-gateway
+    # appgw.ingress.kubernetes.io/cookie-based-affinity: "true"
+
+    # nginx
+    nginx.ingress.kubernetes.io/rewrite-target: /
```

Comment:
Thank you for the detailed report!

When using AWS ALB, the following default configuration should work,

- `.spec.rules[0].http.paths[0].path` -> `/graphql`
- `.spec.rules[0].http.paths[0].pathType` -> `ImplementationSpecific`

but with Nginx, they needed to be changed as follows.

- `.spec.rules[0].http.paths[0].path` -> `/`
- `.spec.rules[0].http.paths[0].pathType` -> `Prefix`

Sorry about not documenting this while I put the annotations for the Nginx Ingress Controller. Can you give it another try with these settings?

Also, `nginx.ingress.kubernetes.io/rewrite-target` looks unnecessary because we don't "rewrite" paths. However, as you mentioned, after re-investigation I found that the cookie affinity doesn't seem to work without the `rewrite-target` annotation for some reason.

Without `nginx.ingress.kubernetes.io/rewrite-target`, the server doesn't return a `Set-Cookie` header:

```console
$ curl -s -D - http://`minikube ip`/graphql
HTTP/1.1 200 OK
Date: Thu, 19 May 2022 06:01:05 GMT
Content-Type: text/html
Content-Length: 1002
Connection: keep-alive

<!DOCTYPE html>
<html>
...
```

While with `nginx.ingress.kubernetes.io/rewrite-target: /`, the cookie header appears:

```console
$ curl -s -D - http://`minikube ip`/graphql
HTTP/1.1 200 OK
Date: Thu, 19 May 2022 06:41:19 GMT
Content-Type: text/html
Content-Length: 1002
Connection: keep-alive
Set-Cookie: INGRESSCOOKIE=1652942480.8.1576.733112|2167d624b067026961ae4bee6ede2711; Max-Age=300; Path=/; HttpOnly

<!DOCTYPE html>
<html>
...
```

I haven't looked into this issue, but I think the priority is low since there's no plan to use Nginx Ingress Controller yet in production.

Reply:
Thank you for your confirmation!

---

### About NGINX Ingress with new settings

> Sorry about not documenting this while I put the annotations for the Nginx Ingress Controller. Can you give it another try with these settings?

I tried to use new settings that you showed as follows.
As the result, I was able to execute query via NGINX Ingress!

```yaml
scalarDbGraphQlConfiguration:
  contactPoints: "jdbc:postgresql://postgresql:5432/postgres"
  username: "postgres"
  password: "postgres"
  storage: "jdbc"
  namespaces: "s0"
service:
  type: "LoadBalancer"
ingress:
  enabled: true
  className: nginx
  hosts:
    - host: ""
      paths:
        - path: /
          pathType: Prefix
```
```console
$ kubectl get ingress scalardb-graphql -o jsonpath='{.spec.rules[].http.paths}' | jq .
[
  {
    "backend": {
      "service": {
        "name": "scalardb-graphql",
        "port": {
          "number": 8080
        }
      }
    },
    "path": "/",
    "pathType": "Prefix"
  }
]
```
```console
$ curl http://`minikube ip`/graphql \
>   -H 'Content-Type: application/json' \
>   --data '{"query": "query Get @transaction(commit: true) {t0_get(get: {key: {c1: 1, c2: 11}}) {t0 {c3}}}"}'
{"data":{"t0_get":{"t0":{"c3":"aaa"}}},"extensions":{"transaction":{"id":"3bbf1b67-1d45-4676-8492-2c987d4d8548"}}}
```

---

### About `Set-Cookie` header without `nginx.ingress.kubernetes.io/rewrite-target`

> However, as you mentioned, after re-investigation I found that the cookie affinity doesn't seem to work without the `rewrite-target` annotation for some reason.

Thank you for additional investigation!

I confirmed the cookie without `rewrite-target` annotation as follows.

```console
$ kubectl get ingress scalardb-graphql -o jsonpath='{.metadata.annotations}' | jq .
{
  "alb.ingress.kubernetes.io/healthcheck-path": "/graphql?query=%7B__typename%7D",
  "alb.ingress.kubernetes.io/scheme": "internal",
  "alb.ingress.kubernetes.io/target-group-attributes": "stickiness.enabled=true,stickiness.lb_cookie.duration_seconds=60",
  "alb.ingress.kubernetes.io/target-type": "ip",
  "meta.helm.sh/release-name": "scalardb-graphql",
  "meta.helm.sh/release-namespace": "default",
  "nginx.ingress.kubernetes.io/affinity": "cookie",
  "nginx.ingress.kubernetes.io/session-cookie-hash": "sha1",
  "nginx.ingress.kubernetes.io/session-cookie-max-age": "300",
  "nginx.ingress.kubernetes.io/session-cookie-name": "INGRESSCOOKIE",
  "nginx.ingress.kubernetes.io/session-cookie-path": "/"
}
```
```console
$ curl http://`minikube ip`/graphql \
>   -H 'Content-Type: application/json' \
>   --data '{"query": "query Get @transaction(commit: true) {t0_get(get: {key: {c1: 1, c2: 11}}) {t0 {c3}}}"}' \
>   --verbose
*   Trying 192.168.49.2:80...
* TCP_NODELAY set
* Connected to 192.168.49.2 (192.168.49.2) port 80 (#0)
> POST /graphql HTTP/1.1
> Host: 192.168.49.2
> User-Agent: curl/7.68.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 97
>
* upload completely sent off: 97 out of 97 bytes
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Thu, 19 May 2022 08:57:26 GMT
< Content-Type: application/json
< Content-Length: 114
< Connection: keep-alive
< Set-Cookie: INGRESSCOOKIE=1652950647.196.1383.890033|19fac6f9454307ae8dd6c0d0b1df1876; Max-Age=300; Path=/; HttpOnly
<
* Connection #0 to host 192.168.49.2 left intact
{"data":{"t0_get":{"t0":{"c3":"aaa"}}},"extensions":{"transaction":{"id":"45362846-bcef-4a5c-93eb-279eff6a8c53"}}}
```

As the result, it looks like the `Set-Cookie` header is returned from NGINX Ingress without the `rewrite-target` annotation in my local environment.
It's just a guess, but there may be another dependency between `rewrite-target` and `Set-Cookie` or some difference of environment.

However, I agree with the priority is low.
Since we know the configurations that the NGINX Ingress works well and there is a properly comment in the values.yaml, I think that we don't need to investigate the issue of `rewrite-target` and `Set-Cookie` at this time.
If we plan to use NGINX Ingress Controller in production in the future, let's investigate this issue again.